sleap_roots/graphpipeline.py

```diff
@@ -0,0 +1,142 @@
+        # # get_root_angle(pts: np.ndarray, proximal=True, base_ind=0) -> np.ndarray
+        "lateral_angles_proximal": (get_root_angle, [lateral_pts, True, 0]),
+        "lateral_angles_distal": (get_root_angle, [lateral_pts, False, 0]),
+        # get_root_lengths(pts: np.ndarray) -> np.ndarray
+        "lateral_lengths": (get_root_lengths, [lateral_pts]),
+        # get_bases(pts: np.ndarray) -> np.ndarray
+        "lateral_base_pts": (get_bases, [lateral_pts]),
+        # get_tips(pts)
+        "lateral_tip_pts": (get_tips, [lateral_pts]),
+        # get_base_ys(pts: np.ndarray) -> np.ndarray
+        # or just based on primary_base_pt, but the primary_base_pt trait must generate before
+        # "primary_base_pt_y": (get_pt_ys, [data["primary_base_pt"]]),
+        "primary_base_pt_y": (get_base_ys, [primary_pts]),
+        # get_network_solidity(pts: np.ndarray) -> float
+        "network_solidity": (get_network_solidity, [pts_all_array]),
+        # get_network_distribution_ratio(pts: np.ndarray, fraction: float = 2 / 3) -> float
+        "network_distribution_ratio": (get_network_distribution_ratio, [pts_all_array]),
+        # get_network_distribution(pts: np.ndarray, fraction: float = 2 / 3) -> float
+        "network_length_lower": (get_network_distribution, [pts_all_array, 2 / 3]),
+        # get_tip_ys(pts: np.ndarray) -> np.ndarray
+        "primary_tip_pt_y": (get_tip_ys, [primary_pts]),
+        # get_chull_line_lengths(pts: Union[np.ndarray, ConvexHull]) -> np.ndarray
+        "chull_line_lengths": (get_chull_line_lengths, [pts_all_array]),
+        # count_scanline_intersections(pts: np.ndarray, depth: int = 1080, width: int = 2048, n_line: int = 50) -> np.ndarray
+        "scanline_intersection_counts": (
+            count_scanline_intersections,
+            [pts_all_list, 1080, 2048, 50],
+        ),
+        # get_base_xs(pts: np.ndarray) -> np.ndarray
+        "lateral_base_xs": (get_base_xs, [lateral_pts]),
+        # get_base_ys(pts: np.ndarray) -> np.ndarray
+        "lateral_base_ys": (get_base_ys, [lateral_pts]),
+        # get_tip_xs(pts: np.ndarray) -> np.ndarray
+        "lateral_tip_xs": (get_tip_xs, [lateral_pts]),
+        # get_tip_ys(pts: np.ndarray) -> np.ndarray
+        "lateral_tip_ys": (get_tip_ys, [lateral_pts]),
+        # get_base_tip_dist(pts: np.ndarray) -> np.ndarray
+        "primary_base_tip_dist": (get_base_tip_dist, [primary_pts]),
+        # get_base_length(pts: np.ndarray)
+        "base_length": (get_base_length, [lateral_pts]),
+        # get_grav_index(pts: np.ndarray)
+        "grav_index": (get_grav_index, [primary_pts]),
+    }
+
+    dts = get_traits_graph()
+
+    data = {}
+    for trait_name in dts:
+        print("trait name:", trait_name)
+        # outputs = ("primary_base_pt",)
+        outputs = (trait_name,)
+        fn, inputs = trait_map[trait_name]
+        fn_outputs = fn(*[input_trait for input_trait in inputs])
```

Comment:
I was thinking that we'd actually use the keys as input, not the direct values. I'm a little surprised that we can even parametrize everything in terms of those 4 arrays?

Like I was thinking that the inputs map above would have the key names within `data` so we could do something like:

```py
fn, input_names = trait_map[trait_name]
data[trait_name] = fn(*[data[input_trait] for input_trait in inputs])
```

Reply:
(1) For the proximal/distal angle traits, there will be a boolean parameter, for example: `get_root_angle` function will need `proximal=True` to calculate `primary_angle_proximal` trait.
The way to avoid these boolean parameter might be by separating one function to two functions, for example:  get_root_proximal_angle and get_root_distal_angle.
(2) I was trying to organize the `data` dictionary where saves the arguments using this way:
`args = inspect.getfullargspec(get_root_angle).args
print(args)
# ['pts', 'proximal', 'base_ind']`
We can get arguments for each function, but this also need to modify the function argument names, for example, instead of using pts, we will use primary_pts, lateral_pts, or pts_all.